init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
@ngimel: re multiple streams: does this schema to calculate workspace sound right:
- getSharedWorkspace() will store an array of workspaces for each device instead of one. most of the time, only first entry will be allocated. 
- max size will be calculated for each stream separately the same way we do per device now. We will rely on streams dynamic not changing from warmup iterations - if a new stream group is be created, user needs to call find.reset() first and run warmup again.